lib/livebook/web_socket/client_connection.ex

```diff
@@ -9,128 +10,127 @@ defmodule Livebook.WebSocket.ClientConnection do
 
   @timeout 10_000
   @backoff 5_000
+  @no_state :no_state
+  @loop_ping_delay 5_000
 
   defstruct [:url, :listener, :headers, :http_conn, :websocket, :ref, id: 0, reply: %{}]
 
   @doc """
   Starts a new WebSocket connection with given URL and headers.
   """
-  @spec start_link(pid(), String.t(), Mint.Types.headers()) :: GenServer.on_start()
+  @spec start_link(pid(), String.t(), Mint.Types.headers()) :: :gen_statem.start_ret()
   def start_link(listener, url, headers \\ []) do
-    Connection.start_link(__MODULE__, {listener, url, headers})
+    :gen_statem.start_link(__MODULE__, {listener, url, headers}, [])
   end
 
   @doc """
   Sends a Request to given WebSocket Server.
   """
   @spec send_request(pid(), WebSocket.proto()) :: {atom(), term()}
   def send_request(conn, %_struct{} = data) do
-    Connection.call(conn, {:request, data}, @timeout)
+    :gen_statem.call(conn, {:request, data}, @timeout)
   end
 
-  ## Connection callbacks
+  ## gen_statem callbacks
+
+  @impl true
+  def callback_mode, do: :handle_event_function
 
   @impl true
   def init({listener, url, headers}) do
-    state = %__MODULE__{listener: listener, url: url, headers: headers}
-    {:connect, :init, state}
+    data = %__MODULE__{listener: listener, url: url, headers: headers}
+    {:ok, @no_state, data, {:next_event, :internal, :connect}}
   end
 
   @impl true
-  def connect(_, state) do
-    case Client.connect(state.url, state.headers) do
+  def handle_event(event_type, event_data, state, data)
+
+  def handle_event(:internal, :connect, @no_state, %__MODULE__{} = data) do
+    case Client.connect(data.url, data.headers) do
       {:ok, conn, websocket, ref} ->
-        send(state.listener, {:connect, :ok, :connected})
+        send(data.listener, {:connect, :ok, :connected})
         send(self(), {:loop_ping, ref})
```

Comment:
We can replace this with an internal event, but probably better to take this in smaller steps.